src/portal-impl.c

```diff
@@ -548,21 +548,106 @@ warn_please_use_portals_conf (void)
                   "configuration file");
 }
 
+static PortalImplementation *
+find_fallback_portal_implementation (const char *interface)
+{
+  /* x-d-p-gtk has historically been the portal UI backend used by desktop
+   * environments with no backend of their own.
+   * If it isn't installed, that is not an error: we just don't use it. */
+  for (const GList *l = implementations; l != NULL; l = l->next)
+    {
+      PortalImplementation *impl = l->data;
+
+      if (!g_str_equal (impl->dbus_name, "org.freedesktop.impl.portal.desktop.gtk"))
+        continue;
+
+      if (!g_strv_contains ((const char **)impl->interfaces, interface))
+        continue;
+
+      g_warning ("Choosing %s.portal for %s as a last-resort fallback",
+                 impl->source, interface);
+      warn_please_use_portals_conf ();
+      return impl;
+    }
+
+  return NULL;
+}
+
+static PortalImplementation *
+find_any_portal_implementation (const char *interface)
+{
+  for (const GList *l = implementations; l != NULL; l = l->next)
+    {
+      PortalImplementation *impl = l->data;
+
+      if (!g_strv_contains ((const char **)impl->interfaces, interface))
+        continue;
+
+      g_debug ("Falling back to %s.portal for %s", impl->source, interface);
+      return impl;
+    }
+
+  return find_fallback_portal_implementation (interface);
+}
+
+static PortalImplementation *
+find_portal_implementation_iface (const PortalInterface *iface)
+{
+  for (char *const *portal = iface->portals; *portal; portal++)
+    {
+      g_debug ("Found '%s' in configuration for %s", *portal, iface->dbus_name);
+
+      if (g_str_equal (*portal, "none"))
+        return NULL;
+
+      if (g_str_equal (*portal, "*"))
+        return find_any_portal_implementation (iface->dbus_name);
+
+      for (const GList *l = implementations; l != NULL; l = l->next)
+        {
+          PortalImplementation *impl = l->data;
+
+          if (g_str_equal (*portal, impl->source))
+            return impl;
+        }
+
+      g_debug ("%s.portal is unrecognized", *portal);
+    }
+
+  return find_any_portal_implementation (iface->dbus_name);
+}
+
 PortalImplementation *
 find_portal_implementation (const char *interface)
 {
+  PortalImplementation *impl = NULL;
```

Comment:
I didn't notice this in the first round - any particular reason to be moving this variable here? I think I'd prefer if it's redeclared inside each scope its used